Add ISSU capability and fix workflow bug for nxos_install_os module

mikewiebe opened the pull request:
##### SUMMARY
This update brings 3 major changes to the module.

1. Adds ISSU support for NXOS platforms that are ISSU capable.
    * The module uses the `show install all impact` command to make this determination.
2. Fixes the workflow such that the module no longer errors out each time forcing the use of a block/rescue.
    * Previously the module would always report a failure  due to connection timeout for the install itself and additional tasks were used to check success/failure of the upgrade.
3. Addition of an options `opts` dictionary argument to `load_config` so that we can pass an `ignore_timeout` flag from the `nxos_install_os` module.  This is needed to achieve number 2 above.
    * This argument defaults to `None` so there is no impact to modules that are not using this option.
    * In the future `return_error` could also be passed in using `opts` and any additional future requirements can use `opts` but this is beyond the scope of this PR.

##### ISSUE TYPE
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
nxos_install_os

##### ANSIBLE VERSION
```
ansible 2.5.0 (rel150/issu d541da73ac) last updated 2017/11/03 08:34:08 (GMT -400)
  config file = None
  configured module search path = [u'/Users/mwiebe/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/mwiebe/Projects/nxos_ansible/fix_ansible/lib/ansible
  executable location = /Users/mwiebe/Projects/nxos_ansible/fix_ansible/bin/ansible
  python version = 2.7.13 (default, Apr  4 2017, 08:47:57) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]
```


##### ADDITIONAL INFORMATION
The following describes the high level workflow for each upgrade.

1. Perform module parameter checks.
    * Make sure provider timeout value is 500 seconds or higher and exit if not.
    * Gather platform type information and make this available for use if needed in the module.
    * Determine if kickstart image is needed and exit if not.
2. Gather pre-upgrade information.
    * Gather information from the `show install all impact` command and use this information to determine if a `non-disruptive ISSU capable` upgrade is possible or drop back to disruptive if not.
        * The issu module option controls this workflow by indicating the expected result  for ISSU (required, desired, no)
    * If an unrecoverable error is encountered using the `show install all impact` command the module drops back to a legacy method of collecting similar information.
3. Start actual upgrade/downgrade.
    * Return data gathered in step 2 if any of the following are true.
       * Check mode is enabled
       * An error was encountered.
       * No upgrade is needed because the switch is already running the desired software.
    * Use the `install all` command to upgrade the device and return the data.


NOTE: Tests are not included in this PR because upgrades by nature are disruptive and increase the run times for integration tests significantly.  Tests are being added to our private test harness for regular validation.

mikewiebe committed:
Initial issu support

mikewiebe committed:
Enhance ISSU support

mikewiebe committed:
Additional refactoring to simplify code flow

mikewiebe committed:
Remove debugs and bug fixes

mikewiebe committed:
Update doc example output

mikewiebe committed:
Update provider line in task example

mikewiebe committed:
Remove unneeded else clause and comments

mikewiebe committed:
Fix ansible-doc errors

mikewiebe committed:
Satisfy ansibot requirements

mikewiebe commented on lib/ansible/modules/network/nxos/nxos_install_os.py:
```diff
@@ -25,27 +25,29 @@
-      device has completed all checks and it started to reboot. The second
-      task is needed to wait for the device to come back up. The last two tasks
-      are used to verify the installation process was successful.
+    - Tested against the following platforms and images
```
From offline review comments:
* I need to add additional versions for platforms validated in this documentation.

mikewiebe replied:
Complete.

mikewiebe commented on lib/ansible/modules/network/nxos/nxos_install_os.py:
```diff
@@ -25,27 +25,29 @@
-    - In check mode, the module tells you if the current boot images are set
-      to the desired images.
+    - In check mode, the module will indicate if an upgrade is needed and
+      weather or not the upgrade is disruptive or non-disruptive(ISSU).
```
`/weather/whether/`

mikewiebe replied:
Done

trishnaguha commented on lib/ansible/modules/network/nxos/nxos_install_os.py:
```diff
@@ -55,39 +57,56 @@
         description:
             - Name of the system (or combined) image file on flash.
         required: true
+    timeout:
```
Describing `timeout` as an option of this module in the documentation might confuse user since there doc already has `timeout` option from the `provider`/`top` spec . `timeout` description should be in the `notes` section.

mikewiebe replied:
Yeah, I went back and forth on this.  The only reason I included this here is because it's critical to successful usage of this module.  I have the warning if the setting is not correct though so I will move it to the notes section.

mikewiebe replied:
Done.

trishnaguha commented:
needs_contributor

trishnaguha commented on lib/ansible/modules/network/nxos/nxos_install_os.py:
```diff
@@ -96,125 +115,465 @@
+    check_ansible_timer(module)
+
+    global platform
+    platform = get_platform(module)
```
This isn't used anywhere?

mikewiebe replied:
It's not.  I was using it in an earlier version.  I left it because it's extremely likely that it will be used in a future update to handle a corner case that is platform specific.  I would prefer to document this fact in a comment.

trishnaguha replied:
ACk. Just FYI OS discovery can be done after CLICONF is implemented.

mikewiebe committed:
Update header docs

mikewiebe commented:
@trishnaguha I addressed your comments.  I also looked at the Shippable error but I am pretty sure this has nothing to do with my changes.  Looks like an infra problem.

mikewiebe committed:
Update nxos_install_os.py

mikewiebe commented:
@trishnaguha Shippable problem fixed.  Now just waiting on your approval.
